mujoco_warp/_src/collision_primitive_core.py

```diff
@@ -0,0 +1,2158 @@
+# Copyright 2025 The Newton Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from typing import Tuple
+
+import warp as wp
+
+from .math import closest_segment_point
+from .math import closest_segment_to_segment_points
+from .math import normalize_with_norm
+from .math import orthogonals
+from .types import MJ_MINVAL
+
+wp.set_module_options({"enable_backward": False})
+
+
+class vec8f(wp.types.vector(length=8, dtype=wp.float32)):
+  pass
+
+
+class mat43f(wp.types.matrix(shape=(4, 3), dtype=wp.float32)):
+  pass
+
+
+class mat83f(wp.types.matrix(shape=(8, 3), dtype=wp.float32)):
+  pass
+
+
+@wp.func
+def _compute_rotmore(face_idx: int) -> wp.mat33:
+  rotmore = wp.mat33(0.0)
+
+  if face_idx == 0:
+    rotmore[0, 2] = -1.0
+    rotmore[1, 1] = +1.0
+    rotmore[2, 0] = +1.0
+  elif face_idx == 1:
+    rotmore[0, 0] = +1.0
+    rotmore[1, 2] = -1.0
+    rotmore[2, 1] = +1.0
+  elif face_idx == 2:
+    rotmore[0, 0] = +1.0
+    rotmore[1, 1] = +1.0
+    rotmore[2, 2] = +1.0
+  elif face_idx == 3:
+    rotmore[0, 2] = +1.0
+    rotmore[1, 1] = +1.0
+    rotmore[2, 0] = -1.0
+  elif face_idx == 4:
+    rotmore[0, 0] = +1.0
+    rotmore[1, 2] = +1.0
+    rotmore[2, 1] = -1.0
+  elif face_idx == 5:
+    rotmore[0, 0] = -1.0
+    rotmore[1, 1] = +1.0
+    rotmore[2, 2] = -1.0
+
+  return rotmore
+
+
+@wp.func
+def make_tangent(a: wp.vec3) -> wp.vec3:
+  """Generates a tangent vector orthogonal to the input vector."""
+  a = wp.normalize(a)
+  b, c = orthogonals(a)
+  return b
+
+
+@wp.struct
+class ContactPoint:
+  """Represents a contact point with position, normal, tangent, and distance."""
+
+  pos: wp.vec3
+  normal: wp.vec3
+  tangent: wp.vec3
+  dist: float
+
+
+@wp.func
+def contact(pos: wp.vec3, normal: wp.vec3, tangent: wp.vec3, dist: float) -> ContactPoint:
+  """Creates a ContactPoint from position, normal, tangent, and distance."""
+  return ContactPoint(pos=pos, normal=normal, tangent=tangent, dist=dist)
+
+
+@wp.func
+def contact_auto_tangent(pos: wp.vec3, normal: wp.vec3, dist: float) -> ContactPoint:
```

Comment:
not sure that we need this function. i think it is ok to only have `contact` and then construct a tangent vector using `make_tangent` to then provide to `contact`